roslibrust/src/lib.rs

```diff
@@ -338,16 +361,44 @@ impl Client {
         });
     }
 
-    // TODO apparently rosbridge doesn't support this!
-    // TODO get them to! and then implement!
-    // /// Advertises a service and returns a handle to the service server
-    // /// Serivce will be active until the handle is dropped!
-    // pub async fn advertise_service(
-    //     &mut self,
-    //     topic: &str,
-    // ) -> Result<ServiceServer, Box<dyn Error>> {
-    //     unimplemented!()
-    // }
+    /// Advertises a service and returns a handle to the service server
+    /// Service will be active until the handle is dropped!
+    pub async fn advertise_service<T: RosServiceType>(
+        &mut self,
+        topic: &str,
+        server: fn(
+            T::Request,
+        )
+            -> Result<T::Response, Box<dyn std::error::Error + 'static + Send + Sync>>,
+    ) -> RosLibRustResult<()> {
+        let mut writer = self.writer.write().await;
+        writer.advertise_service(topic, T::ROS_SERVICE_NAME).await?;
+
+        // We need to do type erasure and hide the request by wrapping their closure in a generic closure
+        let erased_closure = move |message: &str| -> Result<
+            serde_json::Value,
+            Box<dyn std::error::Error + Send + Sync>,
+        > {
+            // Type erase the incoming type
+            let parsed_msg = serde_json::from_str(message)?;
+            let response = server(parsed_msg)?;
+            // Type erase the outgoing type
+            let response_string = serde_json::json!(response);
+            Ok(response_string)
+        };
+
+        let res = self
+            .services
+            .insert(topic.to_string(), Box::new(erased_closure));
+        if let Some(_previous_server) = res {
+            warn!("Re-registering a server for a pre-existing topic? Are you sure you want to do this");
+            unimplemented!()
```

Comment:
Lol I hope not because this will crash.

Reply:
https://github.com/Carter12s/roslibrust/issues/29